Update glossary (#1145)
opensafely-core/databuilder: GLOSSARY.md

# Glossary

**Backend.**

> The logical representation of a collection of EHR data.
> A Backend is associated with a provider of EHR data, a data store, and a Query Language.
> For example, the TPP Backend is associated with TPP, Microsoft SQL Server, and Transact-SQL.

Reviewer: ```suggestion
A class which translates between the table structure we present to the user (the ehrQL tables) and the table structure in an EHR provider's database (the database tables).

Each Backend is specific to an EHR provider and a Query Engine, which tells Data Builder how to talk to the kind of database used by the EHR provider. For instance the TPP backend uses the MSSQLQueryEngine which knows how to connect to Microsoft SQL Server and speak Transact-SQL to it.

A Backend is not required in order to connect to a database. If the structure of the database tables exactly matches the structure of the ehrQL tables then there is no translation to be done. In that case, we can query the database just be specifying the appropriate Query Engine.
```

**Contract.**
A specification for a collection of EHR data.

> A Contract is fulfilled by a Table.

Reviewer: We don't _really_ have contracts as entity any more. Obviously there is a `contracts/` directory, but all it's got in it are tables just like those in `tables/`.

I think the idea is that a contract would be something like "a table definition which is implemented by more than one EHR provider". But I don't we've fully settled what we're doing here yet.

**Dataset Definition.**
Describes, using ehrQL, the EHR data to extract from a Backend.

**Domain-Specific Language (DSL).**
A programming language designed for a specific application domain.
For example, ehrQL is a DSL;
it is a programming language designed for the specific application domain of EHR data.

**Dummy Data.**

> Randomly-generated data used as a substitute for the EHR data to extract from a Backend.

Reviewer: ```suggestion
Randomly-generated data used as a substitute for the EHR data when generating a dataset from a dataset definition.
```

**ehrQL.**
The Electronic Health Records Query Language, pronounced *Hercule*.
A Domain-Specific Language (DSL) for EHR data.
It consists of the classes, methods, and functions that are used in Dataset Definitions.

**Electronic Health Records (EHR) data.**
Data about the health of the patients in a population, stored electronically (i.e. digitally).

**Measure.**
A quotient (a numerator divided by a denominator) for a given time interval and period (e.g. each month for a year).

**Query Engine.**
Translates the Query Model into the Query Language associated with a Backend.

> For example, a Query Engine translates the Query Model into Transact-SQL, which is associated with the TPP Backend.

Reviewer: Takes a query (a graph of connected Query Model objects), connects to a database and returns the corresponding data. Usually this will be a relational database speaking some dialect of SQL. But it doesn't have to be.

**Query Language.**
A programming language designed for querying a data store.
For example, ehrQL.

**Query Model.**

> It consists of the objects that are created by ehrQL and passed to the Query Engines.

Reviewer: ```suggestion
The core data structure by which Data Builder represents a query. It is responsible for defining the semantics of all the operations we support on the data, and for ensuring that only semantically valid queries can be constructed. When a user writes ehrQL this constructs a graph of Query Model objects which are then passed to a Query Engine for execution. This provides a clear boundary between the syntactic surface of ehrQL (where expressiveness and user convenience are paramount) and the underlying semantics.
```

**Table.**
The conceptual representation of a collection of EHR data.

> A Table fulfils a Contract.

Reviewer: I think it's worth distinguishing two kinds of table here:

**ehrQL table**
Describes the structure and types of some EHR data. User's can write queries against this data using ehrQL.

**Database table**
Table in the SQL sense as stored in the database of an EHR provider. These are not directly exposed to users. Users interact with ehrQL tables and their queries are translated into queries against database tables by a Backend.
